### DISCUSSION TITLE
Does minio support s3-crt client?

### DISCUSSION THREAD

**Discussion by pnorbert** [minio/minio#16345] [Q&A]
**Is your feature request related to a problem? Please describe.**
I am trying to use the S3CrtClient from the AWS SDK instead of the S3Client because it works faster with a physically remote Ceph  storage. However I also want to talk to a local Minio server as well. The [s3-crt-demo.cpp](https://github.com/awsdocs/aws-doc-sdk-examples/blob/main/cpp/example_code/s3-crt/s3-crt-demo.cpp) example can create a new bucket in my local minio but cannot upload an object there. In my own tries, it cannot download an existing object either.

**Describe the solution you'd like**
Is there a plan to support S3CrtClient in the future? 

**Describe alternatives you've considered**
Basically, I am looking for an SDK that supports various S3 compatible storage servers. I haven't tried Minio SDK yet but have a working implementation with AWS S3Client. The S3CrtClient is much faster when downloading a big chunk of data from a physically remote Ceph storage so I would like to use that but I cannot test it locally with my Minio server.

**Additional context**
As with the S3 client, I had to disable virtual addressing and to override the endpoint in the example to make it talk to Minio/Ceph.

**Comment by klauspost** [comment]
What errors are you seeing?

My guess is that you need to enable path-style URLs. But without more information I am just guessing.

**Comment by pnorbert** [comment]
[aws_sdk_2023-01-03-12.log](https://github.com/minio/minio/files/10337444/aws_sdk_2023-01-03-12.log)

**Comment by pnorbert** [comment]
Sorry, I don't know what do you mean by the path-style. Is this the same settings as with virtual addressing on/off?
Maybe you can guess what goes wrong from the aws log?

**Reply by pnorbert** [reply]
Comparing the log to a successful run to Ceph, I see that it was not even trying to upload anything. It failed from "Acquire connection".
I am running Minio server and the example under WSL2 on Windows. 
Do you happen to know how to set up WSL2 so that getaddrinfo does not fail for the local IP address?

**Reply by pnorbert** [reply]
```
$ socket_getaddrinfo 172.17.128.186:9000
Cannot getaddrinfo() - Name or service not known

$ socket_getaddrinfo 172.17.128.186
Resolved host '172.17.128.186', service '0'

socket(AF_INET , SOCK_STREAM, IPPROTO_TCP) + '172.17.128.186:0'
socket(AF_INET , SOCK_DGRAM , IPPROTO_UDP) + '172.17.128.186:0'
socket(AF_INET , SOCK_RAW   , IPPROTO_IP ) + '172.17.128.186:0'
```

Is there another way to address the Minio server than setting the endpoint with IP:PORT?

**Reply by ravindk89** [reply]
> I am running Minio server and the example under WSL2 on Windows.

We can't really debug WSL2 network stack issues - that's a WSL2 thing, not necessarily a MinIO thing. If you could provide the MinIO startup command + any logs from after startup, that would help a little bit at least.

We do have a Windows-compiled MinIO binary you could use. Is there a reason you cannot use that? Or is your entire stack being developed in WSL2 w/ Windows as your primary OS?

**Reply by pnorbert** [reply]
```
$ MINIO_ROOT_USER=xxxxx MINIO_ROOT_PASSWORD=xxxxxx minio server /opt/minio/data --console-address ":9001"
Warning: Default parity set to 0. This can lead to data loss.
MinIO Object Storage Server
Copyright: 2015-2022 MinIO, Inc.
License: GNU AGPLv3 <https://www.gnu.org/licenses/agpl-3.0.html>
Version: RELEASE.2022-11-11T03-44-20Z (go1.19.3 linux/amd64)

Status:         1 Online, 0 Offline.
API: http://172.17.128.186:9000  http://127.0.0.1:9000
RootUser: adios
RootPass: adios000
Console: http://172.17.128.186:9001 http://127.0.0.1:9001
RootUser: adios
RootPass: adios000

Command-line: https://min.io/docs/minio/linux/reference/minio-mc.html#quickstart
   $ mc alias set myminio http://172.17.128.186:9000 adios adios000

Documentation: https://min.io/docs/minio/linux/index.html

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ You are running an older version of MinIO released 1 month ago ┃
┃ Update: Run `mc admin update`                                  ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
```

**Reply by pnorbert** [reply]
I usually cannot see anything running under Windows from WSL2. I am ignorant about networking.

**Reply by pnorbert** [reply]
The trouble is that the AWS S3Crt code works on URLs, and uses getaddrinfo() to get the IP address of the endpoint. This function fails if ports are included. The AWS S3 client does not do this apparently and just uses the (overridden) endpoint to connect to and succeeds. So if anyone is using S3Crt client with Minio, there should be some trick to get these two working together (set the port somewhere else not in endpoint, or set Minio server port to some default value).

**Reply by harshavardhana** [reply]
you can start MinIO at standard ports such as `port 80` @pnorbert

**Reply by pnorbert** [reply]
Thanks for the idea. Indeed I can run it, but only with sudo. It seems to work properly:
```
$ mc alias set myminio http://172.17.128.186
Enter Access Key: 
Enter Secret Key:
Added `myminio` successfully.

$ mc mb myminio/testbucket
Bucket created successfully `myminio/testbucket`.

$ mc cp ny.json myminio/testbucket
...le_code/s3-crt/build/ny.json: 557.97 MiB / 557.97 MiB ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 858.66 MiB/s 0s

$ mc ls myminio/testbucket
[2023-01-03 15:21:40 EST] 558MiB STANDARD ny.json
```

The S3Crt example code however still does not work. It improved, it passed the getaddrinfo() but then it started talking https (port 443) despite I set the endpoint to http://172.17.128.186.

```
[DEBUG] 2023-01-03 20:39:43.176 connection-manager [140428250707712] id=0x55f0f4782880: Acquire connection
[DEBUG] 2023-01-03 20:39:43.176 connection-manager [140428250707712] id=0x55f0f4782880: snapshot - state=1, idle_connection_count=0, pending_acquire_count=1, pending_settings_count=0, pending_connect_count=1, vended_connection_count=0, open_connection_count=0, ref_count=1
[INFO] 2023-01-03 20:39:43.176 connection-manager [140428250707712] id=0x55f0f4782880: Requesting 1 new connections from http
[DEBUG] 2023-01-03 20:39:43.176 dns [140428250707712] id=0x55f0f45a6520: Host resolution requested for 172.17.128.186
[DEBUG] 2023-01-03 20:39:43.176 dns [140428250707712] id=0x55f0f45a6520: cached entries found for 172.17.128.186 returning to caller.
[DEBUG] 2023-01-03 20:39:43.176 task-scheduler [140428217136896] id=0x7fb7f0002560: Scheduling attempt_connection task for immediate execution
[DEBUG] 2023-01-03 20:39:43.176 task-scheduler [140428217136896] id=0x7fb7f0002560: Running attempt_connection task with <Running> status
[DEBUG] 2023-01-03 20:39:43.176 socket [140428217136896] id=0x7fb7e0001600 fd=42: initializing with domain 0 and type 0
[DEBUG] 2023-01-03 20:39:43.176 socket [140428217136896] id=0x7fb7e0001600 fd=42: setting socket options to: keep-alive 1, keep idle 30, keep-alive interval 30, keep-alive probe count 0.
[DEBUG] 2023-01-03 20:39:43.176 socket [140428217136896] id=0x7fb7e0001600 fd=42: beginning connect.
[DEBUG] 2023-01-03 20:39:43.176 socket [140428217136896] id=0x7fb7e0001600 fd=42: connecting to endpoint 172.17.128.186:443.
[DEBUG] 2023-01-03 20:39:43.176 task-scheduler [140428217136896] id=0x7fb7e0001120: Scheduling (null) task for future execution at time 121731063962032
[ERROR] 2023-01-03 20:39:43.176 socket [140428217136896] id=0x7fb7e0001600 fd=42: connection failure
[DEBUG] 2023-01-03 20:39:43.176 channel-bootstrap [140428217136896] id=0x55f0f45a6a60: client connection on socket 0x7fb7e0001600 completed with error 1047.
[DEBUG] 2023-01-03 20:39:43.176 channel-bootstrap [140428217136896] id=0x55f0f45a6a60: recording bad address 172.17.128.186.
```

So at this point it is clearly not a Minio issue but s3-crt using https. I tried to start Minio server on port 443 and omit the port in the setup for s3-crt to make it work, but then it failed early. 

Is there another option to force s3-crt to stick to http and use the port 80, or is there an option to run Minio with https?

Thanks

**Reply by ravindk89** [reply]
https://min.io/docs/minio/linux/operations/network-encryption.html has guidance on configuring MinIO for TLS.

Outside of that, we cannot provide specific guidance on how to make S3CRT work in any specific way. We appreciate your understanding here

**Comment by pnorbert** [comment]
BTW, this seems to be an issue with S3CrtClient: https://github.com/aws/aws-sdk-cpp/issues/1844 but it seems to be unimportant, so people are either not using it or know how to go around the issue. I don't :-(
